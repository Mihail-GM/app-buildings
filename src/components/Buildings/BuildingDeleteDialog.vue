<template>
	<v-dialog v-model="dialogDelete" max-width="500px">
		<v-card >
			<v-card-title style="background: #000000;">
				<h4 class="white--text"> Delete Building </h4>
			</v-card-title>

			<v-card-subtitle class="py-10 text--primary text-justify">
				This will delete the building, all data for this building. This cannot be undone.
				To confirm you wish to delete all data related to this building, enter the building
				name in the box below, and click CONFIRM.
			</v-card-subtitle>

			<v-card-text>
				<v-text-field
					dense
					v-model="inputName"
					label="Building name"
					outlined
				>
				</v-text-field>
			</v-card-text>

			<v-card-actions>
				<v-spacer></v-spacer>
				<v-btn
					id="cancel-button"
					color="red darken-1"
					text
					@click.native="closeDelete"
				>
					Cancel
				</v-btn>

				<v-btn
					id="confirm-delete-button"
					color="green darken-1"
					text
					@click.native="deleteItemConfirm"
				>
					CONFIRM
				</v-btn>
				<v-spacer></v-spacer>
			</v-card-actions>
		</v-card>
	</v-dialog>
</template>

<script>
	export default {
		name: "BuildingDeleteDialog",

		props: {
			value: {
				Boolean,
				default: false
			},

			editedItemProp: {
				type: Object,
				required: true,
				default: {
					uuid: null,
					id: null,
					name: '',
					area: null,
					location: '',
					type: null,
					img: '',
				}
			},
		},

		data() {
			return {
				editedItem: this.editedItemProp,
				inputName: '',
			}
		},

		methods: {
			deleteItemConfirm() {
				if (this.editedItemProp.name === this.inputName) {
					this.$emit('deleteItemConfirm', this.editedItem);
					this.$store.dispatch('snackbar/showSnackbarMessage', {
						message: 'Building was deleted successfully',
						duration: 4000,
						mode: 'success'
					});
					this.closeDelete();
					return;
				}

				this.$store.dispatch('snackbar/showSnackbarMessage', {
					message: 'To delete building names must mach',
					duration: 6000,
					mode: 'fail'
				});
			},

			closeDelete() {
				this.dialogDelete = false;
			},
		},

		computed: {
			dialogDelete: {
				get() {
					return this.value
				},

				set(value) {
					this.$emit('input', value)
				}
			},
		},
	}
</script>

<style scoped>

</style>