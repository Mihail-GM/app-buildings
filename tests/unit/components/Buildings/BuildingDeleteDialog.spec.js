import Vuetify from 'vuetify'
import Vuex from "vuex";

import {createLocalVue, mount} from '@vue/test-utils'
import {merge} from "lodash";

import BuildingDeleteDialog from "@/components/Buildings/BuildingDeleteDialog";
import buildingsStore from "../../../../src/store/modules/buildings";
import flushPromises from "flush-promises";

describe('BuildingDeleteDialog.vue', () => {
	const localVue = createLocalVue();
	localVue.use(Vuex);
	let vuetify;
	let getters;

	beforeEach(() => {
		vuetify = new Vuetify();
	})

	function createStore(overrides) {
		let currentState = "DATA_LOADED";
		let buildings = [
			{
				area: "55",
				id: "2",
				img: "",
				location: "Sofia",
				name: "Luxury house 3",
				type: "house",
				uuid: "-MukqrOCyJiAZW1BCwI0",
			},
			{
				area: "55",
				id: "2",
				img: "",
				location: "Sofia",
				name: "Luxury house 3",
				type: "house",
				uuid: "-MukqrOCyJiAZW1BCwI0",
			}
		];

		getters = {
			getBuildings: () => currentState,
			getCurrentState: () => buildings
		}

		const defaultStoreConfig = {
			modules: {
				buildings: {
					namespaced: true,
					state: {
						currentState,
						buildings,
					},
					getters: buildingsStore.getters,

				}
			}
		};

		return new Vuex.Store(merge(defaultStoreConfig, overrides));
	}

	const createWrapper = (overrides) => {
		const defaultMountingOptions = {
			localVue,
			vuetify,
			store: createStore(),

			mocks: {
				// Vue Router
				$router: {
					push: () => {
					}
				},
			},
		};

		const propsData = {
			value: false,

			editedItemProp: {
				uuid: null,
				id: null,
				name: '',
				area: null,
				location: '',
				type: null,
				img: '',
			}
		}

		return mount(BuildingDeleteDialog, merge(defaultMountingOptions, propsData, overrides))
	}

	it('should render component ', () => {
		const wrapper = createWrapper();

		expect(wrapper.is(BuildingDeleteDialog)).toBe(true);
	})

	it('should update computed value when prop value changed', async () => {
		const wrapper = createWrapper();
		await wrapper.setProps({value: true});

		expect(wrapper.vm.dialogDelete).toBe(true);
	})

	it('should emit input with false value when click cancel button', async () => {
		const closeDeleteChanged = jest.spyOn(BuildingDeleteDialog.methods, "closeDelete");

		const wrapper = createWrapper();
		await wrapper.setProps({value: true});

		let closeButton = wrapper.find('#cancel-button');
		closeButton.trigger('click.stop');

		await flushPromises();
		expect(closeDeleteChanged).toHaveBeenCalled();
		expect(wrapper.emitted().input[0]).toEqual([false]);
	})

	it('should emit input with false value when confirm delete', async () => {
		const wrapper = createWrapper();
		await wrapper.setProps({value: true});

		let confirmButton = wrapper.find('#confirm-delete-button');
		confirmButton.trigger('click.stop');

		await flushPromises();

		expect(wrapper.emitted().input[0]).toEqual([false]);
	})

	it('should match snapshot', () => {
		const wrapper = createWrapper()

		expect(wrapper.html()).toMatchSnapshot()
	})
})
