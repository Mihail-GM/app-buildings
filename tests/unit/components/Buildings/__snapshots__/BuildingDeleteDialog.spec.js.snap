// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BuildingDeleteDialog.vue should match snapshot 1`] = `
<div class="v-dialog__container">
  <!---->
</div>
`;
